modules/class-swsales-module-edd.php

```diff
@@ -0,0 +1,767 @@
+		if ( empty( $sitewide_sale_id ) ) {
+			echo json_encode(
+				array(
+					'status' => 'error',
+					'error'  => esc_html__(
+						'No sitewide sale ID given. Try doing it manually.',
+						'sitewide-sales'
+					),
+				)
+			);
+			exit;
+		}
+
+		/**
+		 * Create a coupon programatically
+		 */
+		global $wpdb;
+		while ( empty( $coupon_code ) ) {
+			$scramble = md5( AUTH_KEY . current_time( 'timestamp' ) . SECURE_AUTH_KEY );
+			$coupon_code = strtoupper( substr( $scramble, 0, 10 ) );			
+		}
+
+		if( !class_exists('EDD_Discount' ) ){
+			return;
+		}
+		
+		$amount        = '50'; // Amount.
+		$discount_type = 'percent'; // Type: flat, percent
+
+		$code = new \EDD_Discount();
+
+		$code->name = $coupon_code;
+		$code->code = $coupon_code;
+		$code->type = $discount_type;
+		$code->amount = $amount;
+
+		$code->save();
+
+		$r = array(
+			'status'      => 'success',
+			'coupon_id'   => $code->ID,
+			'coupon_code' => $coupon_code,
+		);
+		echo wp_json_encode( $r );
+		exit;
+	} // end create_discount_code_ajax()
+
+	/**
+	 * Returns whether the current page is the landing page
+	 * for the passed Sitewide Sale.
+	 *
+	 * @param boolean               $is_checkout_page current value from filter.
+	 * @param SWSales_Sitewide_Sale $sitewide_sale being checked.
+	 * @return boolean
+	 */
+	public static function is_checkout_page( $is_checkout_page, $sitewide_sale ) {		
+		if ( 'edd' !== $sitewide_sale->get_sale_type() ) {
+			return $is_checkout_page;
+		}		
+		return is_page( edd_get_option( 'purchase_page', false ) ) ? true : $is_checkout_page;
+	}
+
+	/**
+	 * Get the coupon for a sitewide sale.
+	 * Callback for the swsales_coupon filter.
+	 */
+	public static function swsales_coupon( $coupon, $sitewide_sale ) {
+		if ( $sitewide_sale->get_sale_type() === 'edd' ) {
+			$coupon_object = new \EDD_Discount( $sitewide_sale->swsales_edd_coupon_id );			
+			if ( ! empty( $coupon_object ) ) {
+				$coupon = $coupon_object->code;
+			}
+		}
+
+		return $coupon;
+	}
+
+	public static function automatic_coupon_application() {
+		$active_sitewide_sale = classes\SWSales_Sitewide_Sale::get_active_sitewide_sale();
+		if ( null === $active_sitewide_sale || 'edd' !== $active_sitewide_sale->get_sale_type() || ! $active_sitewide_sale->should_apply_automatic_discount() ) {
+			return;
+		}
+
+		//Check if we're on the checkout page - Can't on the init hook though. 
+
+		// if ( !is_page( edd_get_option( 'purchase_page', false ) ) ) {
+		// 	return;
+		// }
+
+		$discount_code_id = intval( $active_sitewide_sale->get_meta_value( 'swsales_edd_coupon_id', null ) );
+
+		if ( 0 === $discount_code_id ) {
+			return;
+		}	
+
+		$code = new \EDD_Discount( $discount_code_id );
+ 		if ( !empty( $code ) ) {
+ 			//Set it in the $_REQUEST and EDD Session as the EDD init runs at priority 0
+ 			$_REQUEST['discount'] = $code->code;		 			
+ 			EDD()->session->set( 'preset_discount', $code->code );
+ 		} 		 		
```

Comment:
We check if $_REQUEST['discount'] is set now first. If not then we apply the code, so that the URL can override if necessary.

Reply:
Still think this may need a tweak. Here is what I'm doing to test:
1. Add EDD item to card during SWSale
2. Go to checkout. Discount code is automatically applied
3. Remove applied discount and enter a different one
4. Refresh page

After this process, the SWSales discount overwrites the code manually entered by the user. In the code before setting the discount, we should first check whether there is a discount already being used and, if so, don't try to overwrite it.